llama_index/query_engine/retry_query_engine.py

```diff
@@ -58,3 +61,65 @@ def _query(self, query_bundle: QueryBundle) -> RESPONSE_TYPE:
     async def _aquery(self, query_bundle: QueryBundle) -> RESPONSE_TYPE:
         """Not supported."""
         return self._query(query_bundle)
+
+
+class RetryGuidelineQueryEngine(BaseQueryEngine):
+    """Does retry with evaluator feedback
+    if query engine fails evaluation.
+
+    Args:
+        query_engine (BaseQueryEngine): A query engine object
+        guideline_evaluator (GuidelineEvaluator): A guideline evaluator object
+        resynthesize_query (bool): Whether to resynthesize query
+        max_retries (int): Maximum number of retries
+        callback_manager (Optional[CallbackManager]): A callback manager object
+    """
+
+    def __init__(
+        self,
+        query_engine: BaseQueryEngine,
+        guideline_evaluator: GuidelineEvaluator,
+        resynthesize_query: bool = False,
+        max_retries: int = 3,
+        callback_manager: Optional[CallbackManager] = None,
+        query_transformer: Optional[FeedbackQueryTransformation] = None,
+    ) -> None:
+        self._query_engine = query_engine
+        self._guideline_evaluator = guideline_evaluator
+        self.max_retries = max_retries
+        self.resynthesize_query = resynthesize_query
+        self.query_transformer = query_transformer or FeedbackQueryTransformation(
+            resynthesize_query=self.resynthesize_query
+        )
+        super().__init__(callback_manager)
+
+    def _query(self, query_bundle: QueryBundle) -> RESPONSE_TYPE:
+        """Answer a query."""
+        response = self._query_engine._query(query_bundle)
+        if self.max_retries <= 0:
+            return response
+        typed_response = (
```

Comment:
nit: related to another comment. It seems like if the interface just takes the string response, it would make this easier?

Reply:
This is based on the self._query_engine._query(query_bundle) interface so I don't think I can change it.

Reply:
Okay no worries.

I think my general comment here is that if `BaseEvaluator.evaluate_response` takes in either `Response` objects or `str` response. It would be a more generally useful module for users.

Reply:
Yeah currently the old functions uses the source nodes in the Response so I can't just replace it. Maybe will require a rewrite later.